### DISCUSSION TITLE
How to support 360° rotate object detection （Object Head Detection） in MMRotate?

### DISCUSSION THREAD

**Discussion by vansin** [open-mmlab/mmrotate#689] [General]
如何在 MMRotate 中支持 360°的旋转目标检测？
How to support 360° rotate target detection in MMRotate?

# 背景动机

现有的 MMRotate 的算法检测出来的旋转框，无论是 OpenCV表示法还是长边表示法，表示的范围只有180°，不能区分旋转框的头部和尾部。而头部和尾部信息在现实生活中是非常重要，例如区分出遥感图像中车辆的头部和尾部，飞机的头部和尾部时极具价值的。所以希望能够在 MMRotate 中实现 360°的旋转目标检测。

# 相关论文

(以下论文来自 yangxue 大佬的推荐)
1. （比较推荐）根据头部定义框，框的角度定义范围就是0-360，然后iou、nms以及不同框的转换函数对应改一下就行（iou应该直接用都不用改），参考文献：
EAGLE: Large-Scale Vehicle Detection Dataset in Real-World Scenarios using Aerial Imagery
DRBox-v2: An improved detector with rotatable boxes for target detection in SAR images

2. 先检测旋转框，然后对四边进行一次分类，分出头部所在的边，参考文献：
On the Arbitrary-Oriented Object Detection: Classification based Approaches Revisited

3. https://github.com/yanqingyao1994/QPDet

# 数据集约定

在本部分，如何编码头部信息比较好？

## DOTA 格式(起点顺时针定义法)

![image](https://user-images.githubusercontent.com/25839884/211377244-3d3cc608-a09d-40d4-8891-479dced738d2.png)

强制约定A为物体的起始点，A、B、C和D点顺时针排布。此时边AB物体的头部，在DOTA格式的数据集中就不必要标出物体的头部了。

# 评测指标

在经典的mAP的基础，类别，iou＞0.5，头部三个条件满足才算TP

# 旋转目标检测头部的可视化

此部分时实现旋转目标检测的头部的可视化~

- 在检测出来的旋转框中标出 A、B、C和D 顶点
- AB边使用特殊颜色进行可视化

**Comment by vansin** [comment]
# 先检测旋转框，然后对四边进行一次分类

如果我想在 rotate-retinanet 上添加头部检测的功能，是不是只要在 retinanet 的头部新增一个头部的分类的编码就行？

**Reply by yangxue0827** [reply]
我是觉得这样对mmrotate的功能扩展是最好最简单的，方便有需要的研究者接着研究。我之前标了一个带有头部（第一个标注点）的的数据集，可以用来实验。包含OHD-SJTU-S和OHD-SJTU-L两个版本。OHD-SJTU-L是在OHD-SJTU-S基础上加入了DOTA-v1.0 tranval部分有头部的类别，DOTA-v1.0头部信息标注不是很明确，比较脏，所以可以只用里面的OHD-SJTU-S数据集。
https://yangxue0827.github.io/OHD-SJTU.html

**Reply by qwert31639** [reply]
如果是要偵測頭部數據，可以考慮從魚眼開始，因為魚眼中的人物有一個越接近攝影機越靠近圖像中心的特性，我們可以根據這個對圖像進行整理
https://vip.bu.edu/projects/vsns/cossy/datasets/wepdtof/download/ (轉貼，但這組數據需要自己進行轉換)
或著帶有skeleton的數據集e.g. coco也可以藉由骨架中的頭腳連線設定旋轉軸

至於設定一組旋轉360的方法，我在1.x的版本中有試著改動過其實只要修改[mmrotate/structures/bbox/RotatedBoxes](https://github.com/open-mmlab/mmrotate/blob/1.x/mmrotate/structures/bbox/rotated_boxes.py)
我主要修改regularize_boxes，corner2rbox，flip_，rotate_ 這四組
以及
https://github.com/open-mmlab/mmrotate/blob/766185ed317f99379cb14035a6f9e5cf8a5340ad/mmrotate/models/task_modules/coders/delta_xywht_rbbox_coder.py#L61
我新的定義角度範圍為

               0
               |
               |
    -90---------90
               |
               |
          +-180
v0.3.3就還要多考慮pipeline的修正

**Reply by yangxue0827** [reply]
如果方便的话，非常欢迎您提PR来完善mmrotate这部分功能。

**Reply by vansin** [reply]
> 如果是要偵測頭部數據，可以考慮從魚眼開始，因為魚眼中的人物有一個越接近攝影機越靠近圖像中心的特性，我們可以根據這個對圖像進行整理 https://vip.bu.edu/projects/vsns/cossy/datasets/wepdtof/download/ (轉貼，但這組數據需要自己進行轉換) 或著帶有skeleton的數據集e.g. coco也可以藉由骨架中的頭腳連線設定旋轉軸
> 
> 至於設定一組旋轉360的方法，我在1.x的版本中有試著改動過其實只要修改[mmrotate/structures/bbox/RotatedBoxes](https://github.com/open-mmlab/mmrotate/blob/1.x/mmrotate/structures/bbox/rotated_boxes.py) 我主要修改regularize_boxes，corner2rbox，flip_，rotate_ 這四組 以及
> 
> https://github.com/open-mmlab/mmrotate/blob/766185ed317f99379cb14035a6f9e5cf8a5340ad/mmrotate/models/task_modules/coders/delta_xywht_rbbox_coder.py#L61
> 
> 
> 我新的定義角度範圍為
> ```
>            0
>            |
>            |
> -90---------90
>            |
>            |
>       +-180
> ```
> 
> v0.3.3就還要多考慮pipeline的修正

您是否有可供参考的开源代码~

**Reply by vansin** [reply]
> 如果是要偵測頭部數據，可以考慮從魚眼開始，因為魚眼中的人物有一個越接近攝影機越靠近圖像中心的特性，我們可以根據這個對圖像進行整理 https://vip.bu.edu/projects/vsns/cossy/datasets/wepdtof/download/ (轉貼，但這組數據需要自己進行轉換) 或著帶有skeleton的數據集e.g. coco也可以藉由骨架中的頭腳連線設定旋轉軸
> 
> 至於設定一組旋轉360的方法，我在1.x的版本中有試著改動過其實只要修改[mmrotate/structures/bbox/RotatedBoxes](https://github.com/open-mmlab/mmrotate/blob/1.x/mmrotate/structures/bbox/rotated_boxes.py) 我主要修改regularize_boxes，corner2rbox，flip_，rotate_ 這四組 以及
> 
> https://github.com/open-mmlab/mmrotate/blob/766185ed317f99379cb14035a6f9e5cf8a5340ad/mmrotate/models/task_modules/coders/delta_xywht_rbbox_coder.py#L61
> 
> 
> 我新的定義角度範圍為
> ```
>            0
>            |
>            |
> -90---------90
>            |
>            |
>       +-180
> ```
> 
> v0.3.3就還要多考慮pipeline的修正

请问您有更新 mAP 的计算方式不哇？ 360 度时您是怎么处理的？

**Reply by vansin** [reply]
> 如果是要偵測頭部數據，可以考慮從魚眼開始，因為魚眼中的人物有一個越接近攝影機越靠近圖像中心的特性，我們可以根據這個對圖像進行整理 https://vip.bu.edu/projects/vsns/cossy/datasets/wepdtof/download/ (轉貼，但這組數據需要自己進行轉換) 或著帶有skeleton的數據集e.g. coco也可以藉由骨架中的頭腳連線設定旋轉軸
> 
> 至於設定一組旋轉360的方法，我在1.x的版本中有試著改動過其實只要修改[mmrotate/structures/bbox/RotatedBoxes](https://github.com/open-mmlab/mmrotate/blob/1.x/mmrotate/structures/bbox/rotated_boxes.py) 我主要修改regularize_boxes，corner2rbox，flip_，rotate_ 這四組 以及
> 
> https://github.com/open-mmlab/mmrotate/blob/766185ed317f99379cb14035a6f9e5cf8a5340ad/mmrotate/models/task_modules/coders/delta_xywht_rbbox_coder.py#L61
> 
> 
> 我新的定義角度範圍為
> ```
>            0
>            |
>            |
> -90---------90
>            |
>            |
>       +-180
> ```
> 
> v0.3.3就還要多考慮pipeline的修正
请问下段代码是否有参考的哇~，好像问得太多了嘿嘿嘿嘿

https://github.com/vansin/mmrotate/blob/427f47d6d346b2db1777b300027f23d9ae66c6f1/mmrotate/structures/bbox/transforms.py#L6

```python
def norm_angle(angle, angle_range):
    """Limit the range of angles.

    Args:
        angle (ndarray): shape(n, ).
        angle_range (Str): angle representations.

    Returns:
        angle (ndarray): shape(n, ).
    """
    if angle_range == 'oc':
        return angle
    elif angle_range == 'le135':
        return (angle + np.pi / 4) % np.pi - np.pi / 4
    elif angle_range == 'le90':
        return (angle + np.pi / 2) % np.pi - np.pi / 2
    else:
        print('Not yet implemented.')

```

**Reply by qwert31639** [reply]
mAP那個不確定您的意思是什麼，不過在我的task中我依然採用DOTAMetric的評估方式，具體設置如下圖
![image](https://user-images.githubusercontent.com/34253456/212774639-123fee30-f045-44cb-8fd5-abd348c9a668.png)
運行時會看到
![image](https://user-images.githubusercontent.com/34253456/212775736-151f0202-b4d9-43d3-bbb7-58a1e7acb4b9.png)

然後pipline的設置會像是這樣
![image](https://user-images.githubusercontent.com/34253456/212774938-d48db222-ccfa-4c64-b92e-e45ee338944c.png)
其實主要作用的函式還是在https://github.com/vansin/mmrotate/blob/fee578afa550ae6579f5505677d58e57bf2d46b5/projects/headet/structures/bbox/rotated_boxes.py#L24
在1.x的版本中pipeline中沒有改動

**Comment by vansin** [comment]
我将 norm_angle 改成 如下，新增  angle_range == 'h180' 的情况，感觉写的不对，没有深刻理解 @qwert31639 , 虽然训练时跑起来了~ 


感觉norm_angle写错了，虽然训练起来了
```python
def norm_angle(angle, angle_range):
    """Limit the range of angles.

    Args:
        angle (ndarray): shape(n, ).
        angle_range (Str): angle representations.

    Returns:
        angle (ndarray): shape(n, ).
    """
    if angle_range == 'oc':
        return angle
    elif angle_range == 'le135':
        return (angle + np.pi / 4) % np.pi - np.pi / 4
    elif angle_range == 'le90':
        return (angle + np.pi / 2) % np.pi - np.pi / 2
    elif angle_range == 'h180':
        return (angle + np.pi) % np.pi - np.pi
    else:
        print('Not yet implemented.')

```
使用了
[projects/headet/configs/rotated_retinanet/rotated-retinanet-rbox-h180_r50_fpn_1x_dota.py](https://github.com/vansin/mmrotate/blob/qd/projects/headet/configs/rotated_retinanet/rotated-retinanet-rbox-h180_r50_fpn_1x_dota.py)

[projects/headet/structures/bbox/rotated_boxes.py#L24](https://github.com/vansin/mmrotate/blob/cb9a107fe4d2f0b739644e5dfc7dc55f99c9ef9f/projects/headet/structures/bbox/rotated_boxes.py#L24)

[mmrotate/structures/bbox/transforms.py#L22](https://github.com/vansin/mmrotate/blob/cb9a107fe4d2f0b739644e5dfc7dc55f99c9ef9f/mmrotate/structures/bbox/transforms.py#L22)

```shell
python tools/train.py projects/headet/configs/rotated_retinanet/rotated-retinanet-rbox-h180_r50_fpn_1x_dota.py
```

![image](https://user-images.githubusercontent.com/25839884/212758281-59a1fe81-4e4d-446f-be1d-9b94b1ab9c9f.png)

**Reply by vansin** [reply]
思考了一下 ,h180的norm_angle应该是

```python
return (angle + np.pi) % (2*np.pi) - np.pi
```

**Reply by qwert31639** [reply]
抱歉我忘記貼norm_angle 相關的code了，其實在我的task中我是這樣寫的
![image](https://user-images.githubusercontent.com/34253456/212774240-86b50ea2-640d-4efb-ae58-a2c308d3d142.png)

**Reply by qwert31639** [reply]
然後回覆一下，如果您需要由+-180轉成360只需要在h180的基礎上更改
`angle = (angle + 2*np.pi)%(2*np.pi)` 即可

**Reply by vansin** [reply]
> 抱歉我忘記貼norm_angle 相關的code了，其實在我的task中我是這樣寫的 ![image](https://user-images.githubusercontent.com/34253456/212774240-86b50ea2-640d-4efb-ae58-a2c308d3d142.png)

没太懂哇~，可能我没太懂下面的定义~

```shell
           0
           |
           |
-90---------90
           |
           |
      +-180

```

**Reply by qwert31639** [reply]
原本1.x的code採用了cv2.minAreaRect，這會讓我找不到原始第一條邊的位置(也許大神有其他的解法)，所以我從新定義flip、rotate等函數，規則就是按照+-180度的邏輯走的，自然後面輸出的角度就不需要正規畫的動作

**Reply by vansin** [reply]
h180 为什么是 

```python
return angle
```

而不是

```python
return (angle + np.pi) % (2*np.pi) - np.pi
```

第二种方式的return 的范围才是 -pi,pi,第一种return角度的范围不是没有限制吗？ 可能我没有上下文

**Reply by vansin** [reply]
> ```shell
>            0
>            |
>            |
> -90---------90
>            |
>            |
>       +-180
> ```
感觉是不是下面这样也可以哇~
```shell
           -90
           |
           |
+-180---------0
           |
           |
      90

```

**Reply by qwert31639** [reply]
的確，這個部份其實我沒有寫得很好，我就只是把正規化的邏輯放在了每一次旋轉或翻轉中比如說
https://github.com/vansin/mmrotate/blob/cb9a107fe4d2f0b739644e5dfc7dc55f99c9ef9f/projects/headet/structures/bbox/rotated_boxes.py#L123
以及
https://github.com/vansin/mmrotate/blob/cb9a107fe4d2f0b739644e5dfc7dc55f99c9ef9f/projects/headet/structures/bbox/rotated_boxes.py#L142
所以後面才可以直接輸出角度，不過你寫的那個也是對的

**Reply by vansin** [reply]
谢谢~

**Comment by vansin** [comment]
角度预测出来之后，什么时候用到以下代码呢？
```python
@register_box_converter(RotatedBoxes, QuadriBoxes, force=True)
def rbox2qbox(boxes: Tensor) -> Tensor:
    """copy from mmrotate/structures/bbox/box_converters.py Convert rotated
    boxes to quadrilateral boxes.

    Args:
        boxes (Tensor): Rotated box tensor with shape of (..., 5).
    Returns:
        Tensor: Quadrilateral box tensor with shape of (..., 8).
    """
    centerx, centery, w, h, theta = torch.split(boxes, (1, 1, 1, 1, 1), dim=-1)
    # cos_value, sin_value = torch.cos(theta), torch.sin(theta)
    cosa = torch.cos(theta)
    sina = torch.sin(theta)
    wx, wy = w / 2 * cosa, w / 2 * sina
    hx, hy = -h / 2 * sina, h / 2 * cosa
    p1x, p1y = centerx - wx - hx, centery - wy - hy
    p2x, p2y = centerx + wx - hx, centery + wy - hy
    p3x, p3y = centerx + wx + hx, centery + wy + hy
    p4x, p4y = centerx - wx + hx, centery - wy + hy
    return torch.stack([p1x, p1y, p2x, p2y, p3x, p3y, p4x, p4y], dim=-1)

```

**Reply by qwert31639** [reply]
這一段是我為了要抓出第一條邊而從新定義的，用在RotLocalVisualizer裡面

**Reply by vansin** [reply]
谢谢~,Visualizer 里面也有修改可视化第一条边的代码是哇~

**Reply by vansin** [reply]
> 這一段是我為了要抓出第一條邊而從新定義的，用在RotLocalVisualizer裡面
Visualizer 这部分代码方不方便分享一下哇~

**Reply by qwert31639** [reply]
其實那一段只是方便我看val的狀況而已，對於訓練沒有太大幫助，實際上也只有多畫了一個點
原始碼
![image](https://user-images.githubusercontent.com/34253456/213422864-2c6ab5d6-1e25-4cc4-97f5-e34262ab735c.png)
大概的效果(並非真實val的樣子)
![image](https://user-images.githubusercontent.com/34253456/213421798-395bdbc1-70eb-40a1-ac0d-3a8b265226e9.png)
原始圖片來自
https://vip.bu.edu/projects/vsns/cossy/datasets/wepdtof/download/

**Reply by vansin** [reply]
thanks

**Reply by vansin** [reply]
> 角度预测出来之后，什么时候用到以下代码呢？
> 
> ```python
> @register_box_converter(RotatedBoxes, QuadriBoxes, force=True)
> def rbox2qbox(boxes: Tensor) -> Tensor:
>     """copy from mmrotate/structures/bbox/box_converters.py Convert rotated
>     boxes to quadrilateral boxes.
> 
>     Args:
>         boxes (Tensor): Rotated box tensor with shape of (..., 5).
>     Returns:
>         Tensor: Quadrilateral box tensor with shape of (..., 8).
>     """
>     centerx, centery, w, h, theta = torch.split(boxes, (1, 1, 1, 1, 1), dim=-1)
>     # cos_value, sin_value = torch.cos(theta), torch.sin(theta)
>     cosa = torch.cos(theta)
>     sina = torch.sin(theta)
>     wx, wy = w / 2 * cosa, w / 2 * sina
>     hx, hy = -h / 2 * sina, h / 2 * cosa
>     p1x, p1y = centerx - wx - hx, centery - wy - hy
>     p2x, p2y = centerx + wx - hx, centery + wy - hy
>     p3x, p3y = centerx + wx + hx, centery + wy + hy
>     p4x, p4y = centerx - wx + hx, centery - wy + hy
>     return torch.stack([p1x, p1y, p2x, p2y, p3x, p3y, p4x, p4y], dim=-1)
> ```

qbox2rbox 的函数是不是也需要重写

**Reply by vansin** [reply]
感觉需要修改一下

```python
@register_box_converter(QuadriBoxes, RotatedBoxes)
def qbox2rbox(boxes: Tensor) -> Tensor:
    """Convert quadrilateral boxes to rotated boxes.

    Args:
        boxes (Tensor): Quadrilateral box tensor with shape of (..., 8).

    Returns:
        Tensor: Rotated box tensor with shape of (..., 5).
    """
    # TODO support tensor-based minAreaRect later
    original_shape = boxes.shape[:-1]
    points = boxes.cpu().numpy().reshape(-1, 4, 2)
    rboxes = []
    for pts in points:
        (x, y), (w, h), angle = cv2.minAreaRect(pts)
        rboxes.append([x, y, w, h, angle / 180 * np.pi])
    rboxes = boxes.new_tensor(rboxes)
    return rboxes.view(*original_shape, 5)

```

**Reply by qwert31639** [reply]
@vansin 目前這個函數沒有使用到，但如果需要改可以參考
 numpy版本
![image](https://user-images.githubusercontent.com/34253456/215367364-556b8824-47dd-426f-a057-296e127f83d0.png)

**Comment by qwert31639** [comment]
剛剛稍微改了一下(還沒寫unit test)，大概是這樣
```python
@register_box_converter(QuadriBoxes,RotatedBoxes, force=True)
def qbox2rbox(boxes: Tensor) -> Tensor:
    """Convert quadrilateral boxes to rotated boxes.

    Args:
        boxes (Tensor): Quadrilateral box tensor with shape of (..., 8).

    Returns:
        Tensor: Rotated box tensor with shape of (..., 5).
    """
   points = boxes.cpu().numpy().reshape(-1, 4, 2)
   cxs    = points[...,0].mean()
   cys    = points[...,1].mean()
   widthes= np.linalg.norm(points[:,1,:] - points[:,0,:]) 
   heights= np.linalg.norm(points[:,2,:] - points[:,1,:])
   thetas = np.arctan2((points[:,1,0]- points[:,2,0]),
                       (points[:,1,1]- points[:,2,1]))
   thetas = np.where(thetas >= 0, np.pi - thetas, -(np.pi + thetas))
   return torch.stack([cxs, cys, widthes, heights, thetas], dim=-1)
```

**Comment by cmyers009** [comment]
Is this still an open issue, or has anyone solved it?

**Comment by lawen5** [comment]
请问这个问题解决了吗
